### PR TITLE
Add BlitImage (Vulkan Only)

### PR DESCRIPTION
**Pull request by jorgeag-google**
The second of a series of CL that integrates the commits from:

https://github.com/bjoeris/bigwheels/tree/experimental-foveation

into BW main's branch.

The series will have 4 CL, this is the second one that integrates the following commit

Add BlitImage (Vulkan Only)

### CODE REVIEW

File: include/ppx/grfx/grfx_command.h [hunk 1]
```diff
@@ -162,6 +162,40 @@ struct ImageToImageCopyInfo
     } extent;
 };
 
+//! @struct ImageBlitInfo
+//!
+//!
+struct ImageBlitInfo
+{
+    struct
```
Comment (wangra-google) [inline include/ppx/grfx/grfx_command.h:170]: srcImage and dskImage use the same struct so that they can be only declared once instead of twice?

**Review by wangra-google — APPROVED** [review]

File: src/ppx/grfx/vk/vk_command.cpp [hunk 2]
```diff
@@ -1035,6 +1035,65 @@ void CommandBuffer::CopyImageToImage(
...
 }
 
+void CommandBuffer::BlitImage(
+    const grfx::ImageBlitInfo* pCopyInfo,
+    grfx::Image*               pSrcImage,
+    grfx::Image*               pDstImage)
+{
+    bool isSourceDepthStencil = grfx::GetFormatDescription(pSrcImage->GetFormat())->aspect == grfx::FORMAT_ASPECT_DEPTH_STENCIL;
+    bool isDestDepthStencil   = grfx::GetFormatDescription(pDstImage->GetFormat())->aspect == grfx::FORMAT_ASPECT_DEPTH_STENCIL;
+    PPX_ASSERT_MSG(isSourceDepthStencil == isDestDepthStencil, "both images in an image copy must be depth-stencil if one is depth-stencil");
```
Comment (apazylbe) [inline src/ppx/grfx/vk/vk_command.cpp:1045]: Is it fine for them to differ if they are not both depth stencil? For example, one depth, one color

**Review by apazylbe — CHANGES_REQUESTED** [review]

File: src/ppx/grfx/vk/vk_command.cpp (hunk 2 above)
Comment (jorgeag-google) [reply src/ppx/grfx/vk/vk_command.cpp:1045]: I belive so, the spec only says that  if one has depth or stencil the other one should also have it. No other restiction on the format

Comment (apazylbe) [reply src/ppx/grfx/vk/vk_command.cpp:1045]: I see from the spec "If either of srcImage or dstImage was created with a depth/stencil format, the other must have exactly the same format", which is not what we are checking now. Also it seems like if formats are one with just depth and one with just stencil the assert will not trigger?

Comment (jorgeag-google) [reply src/ppx/grfx/vk/vk_command.cpp:1045]: Yes, you are right I see your point. LI guess the correct check will be that both (dstImg and srcImg) have the same value in both fields (depth and stencil)

Comment (jorgeag-google) [reply src/ppx/grfx/vk/vk_command.cpp]: Actually, I switched it again to match what I believe the spec says.

File: include/ppx/grfx/dx12/dx12_command.h [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 Google LLC
+// Copyright 2024 Google LLC
```
[the remark this answers is not in the document]
Comment (apazylbe) [reply include/ppx/grfx/dx12/dx12_command.h:1]: Could you revert the copyright year modification?

File: src/ppx/grfx/vk/vk_command.cpp (hunk 2 above)
Comment (apazylbe) [reply src/ppx/grfx/vk/vk_command.cpp]: I think it should be depth or stencil, maybe smth like this `isSourceDepthOrStencil = grfx::GetFormatDescription(pSrcImage->GetFormat())->aspect & grfx::FORMAT_ASPECT_DEPTH_STENCIL;` so it is true if it has either depth or stencil or both
https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/5268c4004524149b6bb7230969038e834914d652/layers/core_checks/cc_copy_blit_resolve.cpp#L2937
Now that I am looking at the other code, I am not sure if this was copied from the other function. By the way, there are a bunch of other rules for using the function, so I am not sure if we need to assert this specific one. Although might still be useful.

File: src/ppx/grfx/vk/vk_command.cpp [hunk 4]
```diff
@@ -1035,6 +1035,67 @@ void CommandBuffer::CopyImageToImage(
         &region);
 }
 
+void CommandBuffer::BlitImage(
+    const grfx::ImageBlitInfo* pCopyInfo,
+    grfx::Image*               pSrcImage,
+    grfx::Image*               pDstImage)
+{
+    bool isSourceDepthStencil = grfx::GetFormatDescription(pSrcImage->GetFormat())->aspect & grfx::FORMAT_ASPECT_DEPTH_STENCIL;
```
Comment (apazylbe) [inline src/ppx/grfx/vk/vk_command.cpp:1043]: isSourceDepthOrStencil

File: src/ppx/grfx/vk/vk_command.cpp [hunk 5]
```diff
@@ -1035,6 +1035,67 @@ void CommandBuffer::CopyImageToImage(
...
+    bool isDestDepthStencil   = grfx::GetFormatDescription(pDstImage->GetFormat())->aspect & grfx::FORMAT_ASPECT_DEPTH_STENCIL;
```
Comment (apazylbe) [inline src/ppx/grfx/vk/vk_command.cpp:1044]: isDestDepthOrStencil

File: src/ppx/grfx/vk/vk_command.cpp [hunk 6]
```diff
@@ -1035,6 +1035,67 @@ void CommandBuffer::CopyImageToImage(
...
+    if (isSourceDepthStencil || isDestDepthStencil) {
+        PPX_ASSERT_MSG(pSrcImage->GetFormat() == pDstImage->GetFormat(), "both images in an image copy must be the same format if one is depth-stencil");
```
Comment (apazylbe) [inline src/ppx/grfx/vk/vk_command.cpp:1046]: "both images in an image copy must have the same format if either has depth or stencil"

File: src/ppx/grfx/vk/vk_command.cpp [hunk 7]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 Google LLC
+// Copyright 2024 Google LLC
```
Comment (apazylbe) [inline src/ppx/grfx/vk/vk_command.cpp:1]: please update back to 2022

File: include/ppx/grfx/vk/vk_command.h [hunk 8]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 Google LLC
+// Copyright 2024 Google LLC
```
Comment (apazylbe) [inline include/ppx/grfx/vk/vk_command.h:1]: please update back to 2022

File: include/ppx/grfx/dx12/dx12_command.h (hunk 3 above)
Comment (apazylbe) [inline include/ppx/grfx/dx12/dx12_command.h:1]: update back to 2022

**Review by apazylbe — COMMENTED** [review]
just a few nits otherwise lgtm

File: src/ppx/grfx/vk/vk_command.cpp (hunk 5 above)
Comment (jorgeag-google) [reply src/ppx/grfx/vk/vk_command.cpp]: Done

**Review by apazylbe — APPROVED** [review]

**Pull request merged**